### ISSUE TITLE
Optionality is not preserved when mapping a mapped type with generics

### ISSUE THREAD

**Issue by aradalvand** [microsoft/TypeScript#43227]
## Summary:
Normally, mapped types preserve optionality, unless specified otherwise, the `Readonly<T>` type is an example:
```typescript
interface A {
    x?: number;
    y: number;
}

type RA = Readonly<A>;
```
The type `RA` above is effectively: `{ readonly x?: number; readonly y: number }`, and so optionality is preserved.

Now, I have a type called `ValidationErrors`, which is a mpped type that transforms a given type to a slightly differnet shape: 

```typescript
type ValidationErrors<T> = {
    [P in keyof T]?: string;
}
```

Below, I declare a variable of type `Readonly<ValidationErrors<T>>` where I specify `T` explicitly, and then I assign an empty object to it, which expectedly yields no errors because `ValidationErrors<T>` makes all properties of `T` optional, and `Readonly<T>` doesn't touch that optionality.
```typescript
let bar: Readonly<ValidationErrors<{ name: string; }>> = {}; // No errors
```

The problem is where I use a generic parameter in place of `T`:

### Actual behavior:
If the type is generic, it complains:
```typescript
function doSomething<T extends {}>() {
    let foo: ValidationErrors<T> = {}; // No errors
    let bar: Readonly<ValidationErrors<T>> = {}; // ERROR
}
```
The error:
```
Type '{}' is not assignable to type 'Readonly<ValidationErrors<T>>'.
```

[Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAaghgGwJYBM7CQewHYFEBO+m+AzgDwAqAfFALxQDeAUFK1ANoAKUS2UA1hBCYAZlAoBdAPwAuKCWD5eAcwDcTAL5MmIgK7YAxhhxQUmAMqYAthGAALFZSgQAHsAjYUJRhqoAKAEpGFjYEWygRTEw5eGQ0YzxCYnJqOh9VKAB6TKgAOUxnJNIQ1jDgKAAjOHw5ACUIODNsBBAyWNR0LESiUkoqGnoGDQzsqFxa2oB5Ws0mIA)

### Expected behavior:
Given the fact that `ValidationErrors<T>` makes all properties optional, and `Readonly<T>` doesn't change the optionality of any properties, the result of `Readonly<ValidationErrors<T>>` must be a type that only contains optional properties, and as such, it must be okay to assign `{}` to it. Therefore, it makes sense that `let foo: ValidationErrors<T> = {}` and `let foo: Readonly<ValidationErrors<T>> = {}` should both work the same way, without yielding any type errors. Am I wrong?

It's also the case even if I do:
```typescript
function doSomething<T extends { name: string }>() {
    let foo: ValidationErrors<T> = { name: 'Jack' }; // No errors
    let bar: Readonly<ValidationErrors<T>> = { name: 'Jack' }; // ERROR
}
```
The error:
```
Type '{ name: string; }' is not assignable to type 'Readonly<ValidationErrors<T>>'.
```

What is the difference between `ValidationErrors<T>` and `Readonly<ValidationErrors<T>>` that makes `{}` assignable to one, but not to the other?

**P.S.** Feel free to close this issue if this IS the intended behavior, but please let me know what the reasoning behind it is, because I'm really confused here and this doesn't make sense to me.

### Technical Details:
- TS Version: 4.1.5
- Editor: VS Code v1.54.2

**Comment by MartinJohns** [comment]
You forgot to fill out the issue template for feature requests / bug reports.

**Comment by AradAral** [comment]
> You forgot to fill out the issue template for feature requests / bug reports.

Is it okay now?

**Comment by MartinJohns** [comment]
Here's a [playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAaghgGwJYBM7CQewHYFEBO+m+AzgDwAqAfFALxQDeAUFK1ANoAKUS2UA1hBCYAZlAoBdAPwAuKCWD5eAcwDcTAL5MmIgK7YAxhhxQUmAMqYAthGAALFZSgQAHsAjYUJRhqoAKAEpGFjYEWygRTEw5eGQ0YzxCYnJqOh9VKAB6TKgAOUxnJNIQ1jDgKAAjOHw5ACUIODNsBBAyWNR0LESiUkoqGnoGDQzsqFxa2oB5Ws0mIA) which you didn't include.

---

I could swear I've seen this issue already, but I can't find it. Probably some way generics are treated, given that they're open types. The TypeScript team will likely share their wisdom.

Just pointing out that `ValidationErrors<Readonly<T>>` works as well.

**Comment by AradAral** [comment]
> Here's a playground link which you didn't include.

Okay, thanks for that.

> I could swear I've seen this issue already, but I can't find it.

I tried searching in the repo but I couldn't find anything related, let me know if you could find it though.

> Just pointing out that ValidationErrors<Readonly<T>> works as well.

That's right, but unfortunately my actual situation is such that I can't use `ValidationError<Readonly<T>>` (because I actually want to have a `DeepReadonly<ValidationResult<T>>`, where `ValidationResult<T>` is a type that has a property of type `ValidationError<T>`, anyways, I don't want to make this post more confusing, but thanks for the pointer.

**Comment by RyanCavanaugh** [comment]
When the target of an assignment is *immediately* a mapped type that optionalizes its properties, we can tell that it's OK to initialize that with `{}`.

When there's an intermediate generic type in the way, though, it's not apparent that `T<Partial<U>>` will necessarily also be all-optional; not all mapped types have this property. When the mapped type *does* have this property, you can always almost rewrite it to `Partial<T<U>>` instead. When the mapped type is more complex (as in `DeepReadonly`), it's beyond our capabilities to conclude that every optional property in will always result *in* an optional property *out*, so we err on the side of disallowing this rather than allowing a possibly-unsound substitution to occur.
